pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentSubscription.java

```diff
@@ -681,5 +862,66 @@ void topicTerminated() {
         }
     }
 
+    /**
+     * Commit a transaction.
+     *
+     * @param txnID         {@link TxnID} to identify the transaction.
+     * @param properties    Additional user-defined properties that can be associated with a particular cursor position.
+     * @throws IllegalArgumentException
+     * @throws TransactionConflictException
+     */
+    public synchronized void commitTxn(TxnID txnID, Map<String,Long> properties) throws TransactionConflictException {
+        if (!this.pendingAckMessagesMap.containsKey(txnID)) {
+            String errorMsg = "[" + topicName + "][" + subName + "] Transaction with id:" + txnID.getMostSigBits()
+                            + txnID.getLeastSigBits() + " not found.";
+            log.error(errorMsg);
+            throw new IllegalArgumentException(errorMsg);
+        }
+
+        // This shouldn't happen.
```

Comment:
After reconsideration, I think this check here doesn't really make sense.
When client commit a transaction, it doesn't necessary to be the ongoing transaction that did cumulative ack on this subscription before.
But we could add timeout check here, so when a transaction did cumulative ack but did commit within time, it's position will be removed. So other subsequent transaction can still use cumulative ack.

Reply:
sgtm